### PR TITLE
Remove dependencies on ContainerInfo

### PR DESCRIPTION
**Pull request by rgulewich**
This is step one in moving to native pods. This attempts to centralize all of the dependencies on ContainerInfo in one place.  Note that this has the nice side effect that we fail early on invalid input: when constructing the container object, rather than in various places around the executor.

There's likely further refactoring that could be done here (eg: the abstraction is still somewhat leaky), but that can be taken on in step two.

### CODE REVIEW

File: executor/runtime/types/types.go [hunk 1]
```diff
@@ -147,51 +125,97 @@ type Container struct {
...
+	TTYEnabled bool
 
 	Config config.Config
 }
 
-func (c *Container) VPCAccountID() string {
-	if vpcAccountID, ok := c.TitusInfo.GetPassthroughAttributes()[AccountIDParam]; ok {
-		return vpcAccountID
-	}
+func (c *Container) GetCapabilities() *titus.ContainerInfo_Capabilities {
```
Comment (sargun) [inline executor/runtime/types/types.go:201]: It'd be nice if all of them were like this.

File: executor/runtime/types/types.go [hunk 2]
```diff
@@ -147,51 +125,97 @@ type Container struct {
...
-	)
-	if details != "" {
-		return fmt.Sprintf("%s-%s-%s", appName, stack, details)
+func combineAppStackDetails(c *Container) string {
+	if c.JobGroupDetail != "" {
+		return fmt.Sprintf("%s-%s-%s", c.AppName, c.JobGroupStack, c.JobGroupDetail)
 	}
-	if stack != "" {
-		return fmt.Sprintf("%s-%s", appName, stack)
+	if c.JobGroupStack != "" {
```
Comment (sargun) [inline executor/runtime/types/types.go:215]: Does it make sense to have `combinedAppStackDetails` as a receiver method of Container?

File: executor/runtime/types/types.go [hunk 3]
```diff
@@ -217,32 +241,32 @@ func (c *Container) GetEnv() map[string]string {
 	env["TITUS_NUM_DISK"] = itoa(c.Resources.Disk)
 	env["TITUS_NUM_NETWORK_BANDWIDTH"] = itoa(c.Resources.Network)
 
-	if name := c.TitusInfo.GetImageName(); name != "" {
+	if name := c.imageName; name != "" {
```
Comment (sargun) [inline executor/runtime/types/types.go:244]: Maybe add a // TODO: Retrieve these from Docker (Pull)?

File: executor/runtime/types/types.go [hunk 4]
```diff
@@ -291,7 +315,7 @@ func (c *Container) GetEnv() map[string]string {
 
 	if c.Config.MetatronEnabled {
 		// When set, the metadata service will return signed identity documents suitable for bootstrapping Metatron
-		env[metadataserverTypes.TitusMetatronVariableName] = "true"
+		env[metadataserverTypes.TitusMetatronVariableName] = True
 	} else {
 		env[metadataserverTypes.TitusMetatronVariableName] = "false"
```
Comment (sargun) [inline executor/runtime/types/types.go:320]: it'd be nice if both True and False were consts

**Review by sargun — APPROVED** [review]

File: executor/runner/runner.go [hunk 5]
```diff
@@ -1,15 +1,13 @@
 package runner
 
 import (
-	"regexp"
-
 	corev1 "k8s.io/api/core/v1"
 
 	"github.com/Netflix/titus-executor/logger"
+	multierror "github.com/hashicorp/go-multierror"
```
[the remark this answers is not in the document]
Comment (rgulewich) [reply executor/runner/runner.go:7]: Wacky how?

[the remark this answers is not in the document]
Comment (sargun) [reply executor/runner/runner.go:7]: I think the newlines are causing the order sorting to be weird.

File: executor/runtime/types/container.go [hunk 6]
```diff
@@ -72,7 +117,36 @@ func NewContainerWithPod(taskID string, titusInfo *titus.ContainerInfo, resource
...
 		Config:             cfg,
 		runtime:            DefaultOciRuntime,
+
+		// Log uploading defaults
+		LogUploadThresholdTime: defaultLogUploadThresholdTime,
+		LogUploadCheckInterval: defaultLogUploadCheckInterval,
+		LogStdioCheckInterval:  defaultStdioLogCheckInterval,
+
+		// Bubbled up fields from ContainerInfo
+		KillWaitSeconds:      titusInfo.GetKillWaitSeconds(),
```
[the remark this answers is not in the document]
Comment (rgulewich) [reply executor/runtime/types/container.go:127]: Do you mean grace period in the pod?  If so, I'm going to move all of the pod-specific logic into its own constructor, and leave the ContainerInfo-specific code separate.  I think mixing and matching is going to make it trickier to reason about.

File: executor/runtime/types/container.go [hunk 7]
```diff
@@ -48,14 +84,23 @@ func itoa(i int64) string {
...
-func NewContainerWithPod(taskID string, titusInfo *titus.ContainerInfo, resources Resources, labels map[string]string, cfg config.Config, pod *corev1.Pod) (*Container, error) {
+func NewContainerWithPod(taskID string, titusInfo *titus.ContainerInfo, resources Resources, cfg config.Config, pod *corev1.Pod) (*Container, error) {
 	networkCfgParams := titusInfo.GetNetworkConfigInfo()
 
+	labels := map[string]string{
+		models.ExecutorPidLabel: fmt.Sprintf("%d", os.Getpid()),
+		models.TaskIDLabel:      taskID,
+	}
+
+	if len(titusInfo.GetIamProfile()) > 0 {
```
[the remark this answers is not in the document]
Comment (rgulewich) [reply executor/runtime/types/container.go:100]: No - I moved this over [from runner.go](https://github.com/Netflix/titus-executor/pull/488/files#diff-82614118e84bda13d86ddbb968d42d6aa0d08e275b6971bb3e498b7b3fa332b0L69). Maybe we can ditch it?

File: executor/runtime/types/types.go [hunk 8]
```diff
@@ -147,51 +125,97 @@ type Container struct {
...
 	// VPC driver fields
-	SecurityGroupIDs   []string
-	Allocation         vpcTypes.HybridAllocation
-	NormalizedENIIndex int
-	BandwidthLimitMbps int64
-	AllocationUUID     string
-
-	// Is this container meant to run SystemD?
-	IsSystemD bool
+	SecurityGroupIDs     []string
```
[the remark this answers is not in the document]
Comment (rgulewich) [reply executor/runtime/types/types.go:128]: Is it fair to say that you'd prefer that the struct is immutable after creation (as much as possible), and so you'd prefer fields be accessed via accessors?

If that's the case, then I think I'd rather turn Container into an interface, and have a version that's backed by ContainerInfo, and one that's backed by the pod (though still parse fields from passthrough attrs when the object is created, so we can error early).  Would that work?

File: executor/runtime/types/types.go [hunk 9]
```diff
@@ -147,51 +125,97 @@ type Container struct {
...
 	pod *corev1.Pod
 
 	// GPU devices
 	gpuInfo GPUContainer
-	// Set if using a non-runc runtime to run system service init commands
-	runtime string
 
-	iamRole string
+	IamRole             string
+	imageName           string
```
[the remark this answers is not in the document]
Comment (rgulewich) [reply executor/runtime/types/types.go:160]: Some of these are actually internal - only accessed indirectly via methods that transform them in some way.

File: executor/runtime/types/types.go (hunk 3 above)
Comment (rgulewich) [reply executor/runtime/types/types.go:244]: Sure - not sure if we'd ever actually implement it, though?

File: executor/runtime/types/types.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (sargun) [reply executor/runtime/types/types.go:128]: IMHO, that would be the "perfect" choice -- even if the pod-based one is initially falling back to containerinfo, but I don't want to do too much refactoring if we do not need to. 

I guess I'd prefer some separation and "knowledge" of fields that can be read directly vs. have to be initialized vs. are mutable.

**Comment by rgulewich** [comment]
@sargun - I made `Container` an interface, then implemented against that interface.  I'd appreciate it if you could take another look, since this is a pretty big change.

**Comment by sargun** [comment]
@rgulewich :+1:

Github added a re-review feature btw

**Review by sargun — APPROVED** [review]

**Pull request merged**